File: android/shared/src/main/kotlin/ch/traintime/shared/api/TrainAPIService.kt
package ch.traintime.shared.api

import ch.traintime.shared.Thresholds
import ch.traintime.shared.models.Departure
import ch.traintime.shared.models.Station
import ch.traintime.shared.models.TransportMode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import org.json.JSONObject
import java.io.IOException
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

sealed class TrainAPIError : Exception() {
    object RateLimited : TrainAPIError()
    data class HttpError(val code: Int) : TrainAPIError()
    object NoData : TrainAPIError()
    object NetworkError : TrainAPIError()
}

data class StationResult(
    val train: List<Station>,
    val bus: List<Station>,
    val tram: List<Station>,
    val special: List<Station>
)

object TrainAPIService {
    private const val BASE_URL = "https://api.traintime.ch"
    private val apiKey: String get() = Secrets.apiKey

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .build()

    suspend fun fetchStations(lat: Double, lon: Double): StationResult = withContext(Dispatchers.IO) {
        val url = "$BASE_URL/v1/nearby?lat=$lat&lon=$lon"
        val json = makeRequest(url)

        StationResult(
            train = parseStationGroup(json.optJSONArray("train"), TransportMode.TRAIN),
            bus = parseStationGroup(json.optJSONArray("bus"), TransportMode.BUS),
            tram = parseStationGroup(json.optJSONArray("tram"), TransportMode.TRAM),
            special = parseStationGroup(json.optJSONArray("special"), TransportMode.SPECIAL)
        )
    }

    suspend fun fetchDepartures(stationId: String): List<Departure> = withContext(Dispatchers.IO) {
        val encoded = URLEncoder.encode(stationId, "UTF-8")
        val url = "$BASE_URL/v1/departures?id=$encoded&limit=${Thresholds.MAX_DEPARTURES}"
        val json = makeRequest(url)

        val array = json.optJSONArray("departures") ?: return@withContext emptyList()
        val deps = mutableListOf<Departure>()
        for (i in 0 until minOf(array.length(), Thresholds.MAX_DEPARTURES)) {
            deps.add(Departure.from(array.getJSONObject(i)))
        }
        deps
    }

    private fun parseStationGroup(array: org.json.JSONArray?, mode: TransportMode): List<Station> {
        if (array == null) return emptyList()
        val stations = mutableListOf<Station>()
        for (i in 0 until array.length()) {
            Station.from(array.getJSONObject(i), mode)?.let { stations.add(it) }
        }
        return stations
    }

    private fun makeRequest(url: String): JSONObject {
        val request = Request.Builder()
            .url(url)
            .addHeader("X-API-Key", apiKey)
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: IOException) {
            throw TrainAPIError.NetworkError
        }

        response.use { resp ->
            if (resp.code == 429) throw TrainAPIError.RateLimited
            if (resp.code != 200) throw TrainAPIError.HttpError(resp.code)
            val body = resp.body?.string() ?: throw TrainAPIError.NoData
            return JSONObject(body)
        }
    }
}

File: apple/TrainTimeWatch/Services/TrainAPIService.swift
import Foundation
import CoreLocation

enum TrainAPIError: Error {
    case rateLimited
    case httpError(Int)
    case noData
    case networkError
}

struct TrainAPIService {
    private static let baseURL = "https://api.traintime.ch"
    private static let apiKey = Secrets.apiKey

    // MARK: - Station Search by Coordinates

    static func fetchStations(
        lat: Double, lon: Double, mode: TransportMode? = nil
    ) async throws -> (train: [Station], bus: [Station], tram: [Station], special: [Station]) {
        var urlString = "\(baseURL)/v1/nearby?lat=\(lat)&lon=\(lon)"
        if let mode = mode {
            switch mode {
            case .bus: urlString += "&mode=bus"
            case .tram: urlString += "&mode=tram"
            case .special: urlString += "&mode=special"
            default: break
            }
        }
        guard let url = URL(string: urlString) else { throw TrainAPIError.noData }

        let (data, response) = try await makeRequest(url: url)
        try checkHTTPResponse(response)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]

        let trainStations = parseStationGroup(json["train"], mode: .train)
        let busStations = parseStationGroup(json["bus"], mode: .bus)
        let tramStations = parseStationGroup(json["tram"], mode: .tram)
        let specialStations = parseStationGroup(json["special"], mode: .special)

        return (trainStations, busStations, tramStations, specialStations)
    }

    // MARK: - Departures

    static func fetchDepartures(stationId: String, favourites: String? = nil) async throws -> (departures: [Departure], favourites: [Departure]) {
        let encoded = stationId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? stationId
        var urlString = "\(baseURL)/v1/departures?id=\(encoded)&limit=\(Thresholds.maxDepartures)"
        if let favParam = favourites?.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) {
            urlString += "&favourites=\(favParam)"
        }
        guard let url = URL(string: urlString) else { throw TrainAPIError.noData }

        let (data, response) = try await makeRequest(url: url)
        try checkHTTPResponse(response)

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        guard let departureArray = json?["departures"] as? [[String: Any]] else {
            return ([], [])
        }

        let deps = departureArray.prefix(Thresholds.maxDepartures).map { Departure.from(json: $0) }
        let favDeps: [Departure]
        if let favArray = json?["favourites"] as? [[String: Any]] {
            favDeps = favArray.map { Departure.from(json: $0) }
        } else {
            favDeps = []
        }
        return (Array(deps), favDeps)
    }

    // MARK: - Formation

    static func fetchFormation(trainNumber: String, date: String, stationId: String, operatorRef: String? = nil) async throws -> Formation? {
        let encodedStation = stationId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? stationId
        var urlString = "\(baseURL)/v1/formation?train=\(trainNumber)&date=\(date)&stop=\(encodedStation)"
        if let op = operatorRef {
            urlString += "&operatorRef=\(op)"
        }
        guard let url = URL(string: urlString) else { return nil }

        let (data, response) = try await makeRequest(url: url)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else { return nil }

        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        return Formation.from(json: json)
    }

    // MARK: - Helpers

    private static func parseStationGroup(_ raw: Any?, mode: TransportMode) -> [Station] {
        guard let array = raw as? [[String: Any]] else { return [] }
        return array.compactMap { Station.from(json: $0, mode: mode) }
    }

    private static func makeRequest(url: URL) async throws -> (Data, URLResponse) {
        var request = URLRequest(url: url)
        request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        do {
            return try await URLSession.shared.data(for: request)
        } catch {
            throw TrainAPIError.networkError
        }
    }

    private static func checkHTTPResponse(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if http.statusCode == 429 { throw TrainAPIError.rateLimited }
        if http.statusCode != 200 { throw TrainAPIError.httpError(http.statusCode) }
    }
}
